server/FuzzServer/sfreerdp.c

```diff
@@ -0,0 +1,937 @@
+
+	if (!(hThread = CreateThread(NULL, 0, test_peer_mainloop, (void*)client, 0, NULL)))
+		return FALSE;
+
+	CloseHandle(hThread);
+	return TRUE;
+}
+
+static DWORD WINAPI test_server_mainloop(void* arg)
+{
+	freerdp_listener* instance = arg;
+
+	WINPR_ASSERT(instance);
+
+	struct server_info* info = instance->info;
+	WINPR_ASSERT(info);
+
+	while (WaitForSingleObject(info->event, 0) != WAIT_OBJECT_0)
+	{
+		HANDLE handles[MAXIMUM_WAIT_OBJECTS] = { 0 };
+		WINPR_ASSERT(instance->GetEventHandles);
+		DWORD count = instance->GetEventHandles(instance, handles, ARRAYSIZE(handles) - 1);
+
+		if (0 == count)
+		{
+			WLog_ERR(TAG, "Failed to get FreeRDP event handles");
+			break;
+		}
+		handles[count++] = info->event;
+
+		const DWORD status = WaitForMultipleObjects(count, handles, FALSE, INFINITE);
+
+		if (WAIT_FAILED == status)
+		{
+			WLog_ERR(TAG, "select failed");
+			break;
+		}
+
+		WINPR_ASSERT(instance->CheckFileDescriptor);
+		if (instance->CheckFileDescriptor(instance) != TRUE)
+		{
+			WLog_ERR(TAG, "Failed to check FreeRDP file descriptor");
+			break;
+		}
+	}
+
+	WINPR_ASSERT(instance->Close);
+	instance->Close(instance);
+	return 0;
+}
+
+static HANDLE run_server(struct server_info* info)
+{
+	WINPR_ASSERT(info);
+
+	WTSRegisterWtsApiFunctionTable(FreeRDP_InitWtsApi());
+	winpr_InitializeSSL(WINPR_SSL_INIT_DEFAULT);
+	freerdp_listener* instance = freerdp_listener_new();
+
+	if (!instance)
+		return NULL;
+
+	instance->info = (void*)info;
+	instance->PeerAccepted = test_peer_accepted;
+
+	WSADATA wsaData = { 0 };
+	if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0)
+		goto fail;
+
+	/* Open the server socket and start listening. */
+	char name[MAX_PATH] = { 0 };
+	UINT16 port = 0;
+	winpr_RAND(&port, sizeof(port));
+
+	sprintf_s(name, sizeof(name), "fuzzsrv.%04" PRIx16, port);
+	info->file = GetKnownSubPath(KNOWN_PATH_TEMP, name);
+
+	if (!info->file)
+		goto fail;
+
+	WINPR_ASSERT(instance->OpenLocal);
+	const BOOL started = instance->OpenLocal(instance, info->file);
+
+	if (!started)
+		goto fail;
+
+	return CreateThread(NULL, 0, test_server_mainloop, instance, 0, NULL);
+
+fail:
+	freerdp_listener_free(instance);
+	WSACleanup();
+	return NULL;
+}
+
+static char* subdir(const char* path, const char* name)
+{
+	char* str = NULL;
+	size_t len = 0;
+	winpr_asprintf(&str, &len, "%s/%s", path, name);
+	return str;
+}
+
+static BOOL generate_certs(struct server_info* info)
+{
+	MAKECERT_CONTEXT* ctx = NULL;
+	BOOL res = FALSE;
+	WINPR_ASSERT(info);
+
+	BYTE random[64] = { 0 };
+	winpr_RAND(random, sizeof(random));
+	char randstr[ARRAYSIZE(random) + 1] = { 0 };
+	for (size_t x = 0; x < ARRAYSIZE(random); x++)
+	{
+		const BYTE val = random[x];
+		const BYTE mod = val % 26;
+		char c = 'a' + mod;
+		randstr[x] = c;
+	}
+	info->path = GetKnownSubPath(KNOWN_PATH_TEMP, randstr);
+	if (!info->path)
+		goto fail;
+	ctx = makecert_context_new();
+	if (!ctx)
+		goto fail;
+
+	char* makecert_argv[] = { "makecert", "-rdp",  "-silent",  "-y",
```

Comment:
I propose to generate certificates one time before a server start.

Reply:
do you have a instruction on how to do that? is there a hook like the `LLVMFuzzerTestOneInput` function?